File: core/java/src/main/java/hara/lang/protocol/IAssoc.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;

@HaraProtocolBinding(namespace = "std.protocol.iassoc", name = "IAssoc")
public interface IAssoc<K, V> {
  @HaraMethod(value = "assoc", arity = 3)
  IAssoc<K, V> assoc(K k, V v);
}

File: core/rust/src/lang/protocol/iassoc.rs
use hara_protocol_macros::hara_protocol;

#[hara_protocol(namespace = "std.protocol.iassoc", name = "IAssoc")]
pub trait IAssoc<K, V> {
    type Output;

    #[hara_method(value = "assoc", arity = 3, whole_wasm)]
    fn assoc(&self, key: K, value: V) -> Self::Output;
}
